### PR TITLE
refactor: rework DB init state machine

### PR DESCRIPTION
**Pull request by crepererum**
Since adding new features like "sequencer replay" or init retries would
make the current code too complex, a refactor is required:

**Config:**
The config struct now holds a `DatabaseState` which is a simple linear
state machine representing the different stages of the database init.

**Init:**
The init module now has a fixpoint-loop which looks at the state,
decides what to do based on it and repeats until either the DB is
initialized or an error occured. This also makes it easier to continue
the init process "in the middle", e.g. when the preserved catalog is
broken or the sequencer (e.g. Kafka) could not be reached.

### CODE REVIEW

File: server/src/init.rs [hunk 1]
```diff
@@ -482,6 +454,95 @@ impl InitStatus {
...
+                InitProgress::Unfinished => {}
+                InitProgress::Done => {
+                    return Ok(true);
+                }
+                InitProgress::Forget => {
+                    return Ok(false);
+                }
+            }
+        }
+    }
```
Comment (crepererum) [inline server/src/init.rs:479]: This is the fixpoint loop.

File: server/src/lib.rs [hunk 2]
```diff
@@ -1918,7 +1928,7 @@ mod tests {
 
         // creating failed DBs does not work
         let err = create_simple_database(&server, "bar").await.unwrap_err();
-        assert_eq!(err.to_string(), "database already exists");
+        assert_eq!(err.to_string(), "database already exists: bar");
```
Comment (crepererum) [inline server/src/lib.rs:1931]: drive-by fix of the "already exists" err msg to also include the DB name

File: server/src/config.rs [hunk 3]
```diff
@@ -374,184 +363,296 @@ impl RemoteTemplate {
 }
 
 #[derive(Debug)]
-struct DatabaseState {
-    db: Arc<Db>,
-    handle: Option<JoinHandle<()>>,
-    shutdown: CancellationToken,
+#[allow(clippy::large_enum_variant)]
+enum DatabaseState {
```
Comment (crepererum) [inline server/src/config.rs:367]: this is the linear state machine that a DB will pass through during init.

File: server/src/config.rs [hunk 4]
```diff
@@ -374,184 +363,296 @@ impl RemoteTemplate {
...
+    },
+
+    /// Rules are loaded
+    RulesLoaded {
+        object_store: Arc<ObjectStore>,
+        exec: Arc<Executor>,
+        server_id: ServerId,
+        rules: DatabaseRules,
+    },
+
```
Comment (crepererum) [inline server/src/config.rs:383]: Soon there will be another state here that captures "catalog loaded but not yet replayed".

**Review by pauldix — APPROVED** [review]
Skimmed through and mostly makes sense to me. I'd like to see a call out to where specific error conditions would get retried vs. just failing out (like some transient object store error while loading something or a hiccup from the write buffer during replay)

**Comment by crepererum** [comment]
> I'd like to see a call out to where specific error conditions would get retried vs. just failing out (like some transient object store error while loading something or a hiccup from the write buffer during replay)

That's something that @mkmik also asked for today. Right now we fail on the first error and the DB will not automatically recover (one exception is that you can order the server to wipe the catalog and it will retry to recover afterwards). That's bad. However with the old code it was pretty hard to factor in retries. With the state-machine + fixpoint approach its way easier (it's just a matter of restarting the loop after a while to retry). This is however NOT done in this PR and I would like to keep it out of it (since writing tests for it is quite some effort and the diff is already quite large).

Does that answer your question?

**Review by mkmik — APPROVED** [review]
